src/Framework/TestCase.php

```diff
@@ -1698,7 +1700,33 @@ private function handleDependencies(): bool
                 }
             }
 
-            $passedKeys = \array_flip(\array_unique($passedKeys));
+            $failedKeys = [];
+
+            foreach ($failed as $failure) {
+                $pos = \strpos($failure->getTestName(), ' with data set');
```

Comment:
The substring ` with [data set] ` is dependent on how test names are send to the loggers. This could change in the future, for example with #3439 

You can retrieve the `TestCase` via `$failure->failedTest()` and explore if it has a data set using `$test->dataName()`

Reply:
It is used like this elsewhere indeed! Which keeps giving me headaches as I am trying to refactor it out. 😆 For the test execution reordering and TestDox original-ordering, I cooked up this one:

https://github.com/sebastianbergmann/phpunit/blob/b8f38d3f22b839ac84c80fc09641de2d8973179f/src/Runner/TestSuiteSorter.php#L77-L94

The ` with data set ` substring is a user interface element that should be independent of the inner workings of the test runner. Here's an upcoming change that might interfere with your code: https://github.com/sebastianbergmann/phpunit/pull/3439/commits/06d8c4b9bff990f18c691e9d70d4240422a91dc8

Reply:
Okey, I'm still not sure how you're gonna replace that "with data set". You would need also refactor these
https://github.com/sebastianbergmann/phpunit/blob/b8f38d3f22b839ac84c80fc09641de2d8973179f/src/Framework/TestCase.php#L1685-L1780

https://github.com/sebastianbergmann/phpunit/blob/b8f38d3f22b839ac84c80fc09641de2d8973179f/src/Framework/TestCase.php#L1097-L1116

I'm still not sure how these should be refactored. 

Also I'm not sure how I can access dependencies of test inside of TestSuiteSorter.

Reply:
There's enough to refactor for everyone! Don't be shy, just grab a few classes and have at it. 🛠 

You can get the name of a `TestCase` with:
- `$testCase->getName($withDataSet)` which gives you the name with(out) a prettified data set
- `$testCase->dataName()` returns the name of the data set you see in `getName()`
- `$testCase->getDependencies()` returns a list of dependencies, if any

Please explore the surrounding code a bit more and write tests _before_ you make such large changes.

Reply:
I understand and I would like to fix things. However I want remind about few facts.

I'm using heavily my own time to fix & learn this stuff. It's not as simple as it looks for guy who have no experience with PHPUnit core. I will look into this after I have created some tests.

Would need this feature ASAP into PHPUnit...

Reply:
I get it, no worries! Like many other tools and frameworks PHPUnit is maintained by volunteer contributors. We are paid in emoji and sleepless nights.

Reply:
@epdenouden Tests are now Implemented :tada:

Reply:
@epdenouden Btw, cannot really remove this `$pos = \strpos($failure->getTestName(), ' with data set');`

Because if taking look into `src/Framework/TestFailure.php` there will be

```php
/**
     * Constructs a TestFailure with the given test and exception.
     *
     * @param Throwable $t
     */
    public function __construct(Test $failedTest, $t)
    {
        if ($failedTest instanceof SelfDescribing) {
            $this->testName = $failedTest->toString();
        } else {
            $this->testName = \get_class($failedTest);
        }

        if (!$failedTest instanceof TestCase || !$failedTest->isInIsolation()) {
            $this->failedTest = $failedTest;
        }

        $this->thrownException = $t;
    }
```

Same with other stuff. Meaning this needs to be refactored and refactoring this means refactoring also skipped, passed, etc... Would change little bit too much logic to be in this PR.